Added TypeScript type definitions

ivanz opened the pull request:
Hi,

This pull request adds type definitions (for TypeScript and editors that support type definitions) and integrates them in package.json. I have added a couple of tests as well to check a couple of things when using the typings.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/astorije/chai-immutable/63)

<!-- Reviewable:end -->

astorije commented:
This is huge, thanks! I'll need some time reviewing this as I have never been nearby TypeScript nor any of the tooling you used here.

I can't promise when, but I'll review this carefully.

astorije commented on chai-immutable.d.ts:
```diff
@@ -0,0 +1,3 @@
+import * as chai from './typing/lib/chai-immutable';
+
+export = chai;
```
Missing EOF here :-)

astorije replied:
On top of the missing EOF, would it make sense to call it `chaiImmutable` or `chaiImmutableTyping` instead of `chai`? (I am still totally inexperienced with TS so just trying to figure out what's in this var 😅)

astorije commented on tslint.json:
```diff
@@ -0,0 +1,7 @@
+  "extends": "tslint-config-typings",
+  "rules": {
+    "no-trailing-whitespace": false,
+    "eofline": false
```
Is this related to previous comment?

ivanz commented:
Review status: 0 of 14 files reviewed at latest revision, 2 unresolved discussions.

---

*[chai-immutable.d.ts, line 3 \[r1\]](https://reviewable.io:443/reviews/astorije/chai-immutable/63#-KJu-q-y2ajI8W8Nz0jA-r1-3:-KJu-z3eaaqmSV73xqk3:1758359711) ([raw file](https://github.com/astorije/chai-immutable/blob/4d3527e10bf7a0128075440678fb403d39a04ae0/chai-immutable.d.ts#L3)):*
<details><summary><i>Previously, astorije (Jérémie Astori) wrote…</i></summary>
> Missing EOF here :-)
</details>
Fixed.

---

*[tslint.json, line 5 \[r1\]](https://reviewable.io:443/reviews/astorije/chai-immutable/63#-KJu-q-y2ajI8W8Nz0jC-r1-5:-KJu0CqLDlBNVs_pSRz0:-785425264) ([raw file](https://github.com/astorije/chai-immutable/blob/4d3527e10bf7a0128075440678fb403d39a04ae0/tslint.json#L5)):*
<details><summary><i>Previously, astorije (Jérémie Astori) wrote…</i></summary>
> Is this related to comment above?
</details>
Yeah - I've changed this so that the whitespace and eof rules are enabled.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/astorije/chai-immutable/63)*
<!-- Sent from Reviewable.io -->

astorije commented:
> Fixed.

You sure? I still see the missing EOF in the files tab.

ivanz commented:
I think GitHub is not displaying them? I can see them locally and tslint passes. If I remove a line it fails. To be honest the new lines don't matter for the typings, because they are just metadata for development-time only (during TypeScript compilation) and it's not something that gets bundled or used during run-time.

astorije commented on .gitignore:
```diff
@@ -1,3 +1,10 @@
+.vscode
```
Can you remove this line from the PR? It's a system-specific ignored file and should be in people's global gitconfig (this also why there is no `.DS_Store` in this file even though I'm running on macOS at the moment).

astorije commented on README.md:
```diff
@@ -311,3 +311,10 @@ Asserts that the immutable collection has the expected size.
+
+## TypeScript type definitions
+
+The `chai-immutable` module contains TypeScript type definitions, which should automatically be picked up by TypeScript.
```
s/The `chai-immutable` module/This plugin/

astorije commented on README.md:
````diff
@@ -311,3 +311,10 @@ Asserts that the immutable collection has the expected size.
 ```
+
+
+## TypeScript type definitions
````
Could you move this section right above the API reference please?

astorije commented on .editorconfig:
```diff
@@ -0,0 +1,5 @@
+
+[*.{js,json}]
+indent_style = space
+indent_size = 2
```
Never used this, wouldn't this file be a simple translation of the lint config file(s)? (I'll move from JSCS to ESLint at some point)

astorije commented on package.json:
```diff
@@ -30,6 +35,12 @@
+      "name": "Ivan Zlatev",
+      "url": "http://ivanz.com"
+    }
+  ],
```
I'm not 100% sure about this, but I do appreciate the help from contributors a lot!
For those who are wondering, I'll add other contributor names once this PR is merged, and if I happen to ever receive hundreds of contributions, I'll rethink this 😄.

astorije commented on package.json:
```diff
@@ -44,6 +55,10 @@
+    "typescript": "^1.8.10",
+    "typings": "^1.0.4",
+    "tslint": "*",
+    "tslint-config-typings": "^0.2.0"
```
That's quite a lot of dependencies added here. Could you add a quick comment about why you use them all for?

Is `typescript` actually necessary in the devDependencies? I'd be annoyed if I had to install an extra 4 dependencies (~40M) to use this small plugin on a JS package.

While we're here, can you bump versions of these dependencies and make sure it still works?

astorije commented on typing/lib/chai-immutable.d.ts:
```diff
@@ -0,0 +1,11 @@
+import {Utils} from 'chai/lib/Utils';
+import {} from './Assertion';
+
+// chai-immutable exports a single function (is a callable? module).
```
Comment needs fixing?

astorije commented on typing/test/Assertation.specs.ts:
```diff
@@ -0,0 +1,30 @@
+import {expect, use} from 'chai';
```
Why are the tests in 2 different files instead of just `chai-immutable.specs.ts`?

astorije commented on package.json:
```diff
@@ -6,13 +6,18 @@
+    "build-typings": "tsc -p .",
+    "test-typings": "npm run build-typings && npm run check-style-typings && mocha typing/build/*.specs.js",
+    "check-style-typings": "tslint typing/lib/**/*.ts typing/test/**/*.ts",
+    "prepublish": "typings install"
```
This will run at each npm publish (i.e. every time I push a tag to GitHub that gets picked up by Travis), is that on purpose?

astorije commented on typings.json:
```diff
@@ -0,0 +1,9 @@
+  "globalDevDependencies": {
+    "mocha": "registry:env/mocha#2.2.5+20160321223601"
+  }
+}
```
This is a bit cryptic to me, what does that mean when I bump dependency versions?
Also, there seems to be a discrepancy between versions here and in `package.json`.

In general, anything in particular I need to know when I add features to the API, publish a version, bump a dep, ...?

ivanz commented:
Hey - I forgot about this pull request :) I am going to close this as it's no longer needed. The new version of Typescript introduced a nice way to pull and manage typings outside of JS libraries and  I saw that @jsebrech has already created some here: https://github.com/jsebrech/chai-immutable-typings

astorije commented:
Argh, I wish I knew that before spending time reviewing :D
Alright, I know nothing about TypeScript, does that mean this plugin doesn't need any change/addition to be TypeScript-friendly?

Does it make sense to add a section in the README, and/or an example?

If I understand properly, you should deprecate https://github.com/ivanz/chai-immutable-typings and redirect to this plugin + https://github.com/jsebrech/chai-immutable-typings or the repo behind https://github.com/DefinitelyTyped/DefinitelyTyped/pull/13067 once it is merged.

ivanz commented:
No changes needed to this repo unless you feel like you want to collaborate with @jsebench to pull his typing file into the repo and publish with the NPM package. Up to you guys.

astorije commented:
Ping @jsebrech, what do you think is the right thing to do here?

jsebrech commented:
You need definition files to pull a non-typescript library into a typescript codebase, but it doesn't matter whether those are included with the base library or distributed separately via definitelytyped / npm. So, this doesn't need to be rolled into chai-immutable, but it could be if someone felt up to maintaining it.

astorije commented:
I don't know the first thing about TypeScript (at least for now) so it might make more sense to keep it in a repo that's managed by TypeScript folks.
However, feel free to open a PR that updates README to document where and how to use these typing files.
